climpred/reference.py

```diff
@@ -0,0 +1,212 @@
+    .. note::
+        Based on Decadal Prediction protocol, this should only be computed for the
+        first lag and then projected out to any further lags being analyzed.
+    Args:
+        hind (xarray object): Initialized ensemble.
+        uninit (xarray object): Uninitialized ensemble.
+        verif (xarray object): Verification data with some temporal overlap with the
+            uninitialized ensemble.
+        metric (str):
+            Metric used in comparing the uninitialized ensemble with the verification
+            data.
+        comparison (str):
+            How to compare the uninitialized ensemble to the verification data:
+                * e2o : ensemble mean to verification data (Default)
+                * m2o : each member to the verification data
+        alignment (str): which inits or verification times should be aligned?
+            - maximize/None: maximize the degrees of freedom by slicing ``hind`` and
+            ``verif`` to a common time frame at each lead.
+            - same_inits: slice to a common init frame prior to computing
+            metric. This philosophy follows the thought that each lead should be based
+            on the same set of initializations.
+            - same_verif: slice to a common/consistent verification time frame prior to
+            computing metric. This philosophy follows the thought that each lead
+            should be based on the same set of verification dates.
+        add_attrs (bool): write climpred compute args to attrs. default: True
+        ** metric_kwargs (dict): additional keywords to be passed to metric
+    Returns:
+        u (xarray object): Results from comparison at the first lag.
+    """
+    # Check that init is int, cftime, or datetime; convert ints or cftime to datetime.
+    hind = convert_to_cftime_index(hind, 'init', 'hind[init]')
+    uninit = convert_to_cftime_index(uninit, 'time', 'uninit[time]')
+    verif = convert_to_cftime_index(verif, 'time', 'verif[time]')
+    has_valid_lead_units(hind)
+
+    # get metric/comparison function name, not the alias
+    metric = METRIC_ALIASES.get(metric, metric)
+    comparison = COMPARISON_ALIASES.get(comparison, comparison)
+
+    comparison = get_comparison_class(comparison, HINDCAST_COMPARISONS)
+    metric = get_metric_class(metric, DETERMINISTIC_HINDCAST_METRICS)
+    forecast, verif = comparison.function(uninit, verif, metric=metric)
+
+    hind = hind.rename({'init': 'time'})
+
+    _, verif_dates = return_inits_and_verif_dates(hind, verif, alignment=alignment)
+
+    plag = []
+    # TODO: Refactor this, getting rid of `compute_uninitialized` completely.
+    # `same_verifs` does not need to go through the loop, since it's a fixed
+    # skill over all leads.
+    for i in hind['lead'].values:
+        # Ensure that the uninitialized reference has all of the
+        # dates for alignment.
+        dates = list(set(forecast['time'].values) & set(verif_dates[i]))
```

Comment:
This might be unsorted after using set if that matters.

Reply:
These are going away anyhow. Just here for the time be.